flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/store/ZooKeeperMesosWorkerStore.java

```diff
@@ -0,0 +1,290 @@
+import org.apache.flink.configuration.ConfigConstants;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.StateHandle;
+import org.apache.flink.runtime.util.ZooKeeperUtils;
+import org.apache.flink.runtime.zookeeper.StateStorageHelper;
+import org.apache.flink.runtime.zookeeper.ZooKeeperStateHandleStore;
+import org.apache.mesos.Protos;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Option;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.ConcurrentModificationException;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A ZooKeeper-backed Mesos worker store.
+ */
+public class ZooKeeperMesosWorkerStore implements MesosWorkerStore {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperMesosWorkerStore.class);
+
+	private final Object cacheLock = new Object();
+
+	/** Client (not a namespace facade) */
+	private final CuratorFramework client;
+
+	/** Flag indicating whether this instance is running. */
+	private boolean isRunning;
+
+	/** A persistent value of the assigned framework ID */
+	private final SharedValue frameworkIdInZooKeeper;
+
+	/** A persistent count of all tasks created, for generating unique IDs */
+	private final SharedCount totalTaskCountInZooKeeper;
+
+	/** A persistent store of serialized workers */
+	private final ZooKeeperStateHandleStore<MesosWorkerStore.Worker> workersInZooKeeper;
+
+	@SuppressWarnings("unchecked")
+	ZooKeeperMesosWorkerStore(
+		CuratorFramework client,
+		String storePath,
+		StateStorageHelper<MesosWorkerStore.Worker> stateStorage
+	) throws Exception {
+		checkNotNull(storePath, "storePath");
+		checkNotNull(stateStorage, "stateStorage");
+
+		// Keep a reference to the original client and not the namespace facade. The namespace
+		// facade cannot be closed.
+		this.client = checkNotNull(client, "client");
+
+		// All operations will have the given path as root
+		client.newNamespaceAwareEnsurePath(storePath).ensure(client.getZookeeperClient());
+		CuratorFramework facade = client.usingNamespace(client.getNamespace() + storePath);
+
+		// Track the assignd framework ID.
+		frameworkIdInZooKeeper = new SharedValue(facade, "/frameworkId", new byte[0]);
+
+		// Keep a count of all tasks created ever, as the basis for a unique ID.
+		totalTaskCountInZooKeeper = new SharedCount(facade, "/count", 0);
+
+		// Keep track of the workers in state handle storage.
+		facade.newNamespaceAwareEnsurePath("/workers").ensure(client.getZookeeperClient());
+		CuratorFramework storeFacade = client.usingNamespace(facade.getNamespace() + "/workers");
+
+		this.workersInZooKeeper = ZooKeeperStateHandleStore.class
+			.getConstructor(CuratorFramework.class, StateStorageHelper.class)
+			.newInstance(storeFacade, stateStorage);
```

Comment:
This is a hack due to the shading of the Curator dependency within flink-runtime.   Other uses of ZooKeeperStateHandleStore occur only within flink-runtime and thus can construct it normally, but to use it in flink-mesos I must use late binding.  Otherwise the compiler complains that its constructor takes something like`org.apache.flink.shaded.org.apache.curator.CuratorFramework`.   

I'll add some commentary.